[Object Detection] Introduce a callback-based coco metric evaluator

ianstenbit opened the pull request:
Example Colab using this:
https://colab.research.google.com/drive/16Lb6rMcFvtoic7DLd89Ap5-VxLa1dX_4?resourcekey=0-sDnh3nGFCx-6c7l_jnxShQ&usp=sharing

Example TensorBoard (from 3 epochs of training a RetinaNet on TPU: https://tensorboard.dev/experiment/Dmvt2HYDTxamopnh5QgJkQ/#scalars&_smoothingWeight=0&runSelectionState=eyJ0cmFpbiI6dHJ1ZX0%3D

Basic API is:

```
from keras_cv.callbacks import PyCOCOCallback
callbacks = [
    PyCOCOCallback(eval_ds, bounding_box_format="xywh"),
    keras.callbacks.TensorBoard(log_dir="logs"),
]

model.fit(
    train_ds,
    validation_data=eval_ds,
    callbacks=callbacks,
)
```

~The evaluation dataset passed to `model.fit` is used for the COCO metrics at epoch end.~ I've updated the API to request its own evaluation dataset, as our call into the validation dataset from the model wasn't stable.
The COCO metrics are created in the logs and should automatically show up in TensorBoard (note that this is currently dependent on the `COCOPyMetric` callback before the `TensorBoard` callback in the callbacks list -- this can be fixed with proper PyMetric support in core Keras)

ianstenbit commented on keras_cv/metrics/coco/pycoco_utils.py:
```diff
@@ -0,0 +1,218 @@
+        return res
+
+
+def _yxyx_to_xywh(boxes):
```
Should just adapt this to use bounding_box.convert_format

bhack replied:
Yes I think it will be useful

bhack commented:
Do we need add a test for this also with a `dummy model` with `compile=True`?

ianstenbit commented:
> Do we need add a test for this also with a `dummy model` with `compile=True`?

Yes absolutely -- this will need tests. I want to get some feedback from @LukeWood and @tanzhenyu on the API surface first

LukeWood commented:
Some feedback:

- lets name the callback COCOEvaluator() or something
- can we support evaluation for the train set from the callback too?

LukeWood commented on keras_cv/metrics/coco/pycoco_utils.py:
```diff
@@ -0,0 +1,231 @@
+
+    coco_eval = cocoeval.COCOeval(coco_gt, coco_dt, iouType="bbox")
+    coco_eval.params.imgIds = image_ids
+    coco_eval.evaluate()
```
does this print a ton of logs?

ianstenbit replied:
Yes it does -- I don't think we can't really do anything about it without forking the Pycocotools code (or with a hack like this: https://stackoverflow.com/questions/8391411/how-to-block-calls-to-print)

LukeWood replied:
unfortunate, but if theres no easy solution it should be fine.

tanzhenyu commented:
One high-level question is : we should probably first provide a VOCEvaluator given our existing training are all on Pascal VOC dataset. That would require some python level coding, but no need to rely on pycoco

innat commented:
@ianstenbit 
Would it be better to introduce `callback` directory?

```python
from keras_cv.metrics import COCOPyMetric # current
from keras_cv.callback import COCOPyMetric # what if
```

Also note, sooner or later, it may be needed introduce **display** type callback for the segmentation or generative model training.

tanzhenyu commented on keras_cv/metrics/__init__.py:
```diff
@@ -12,5 +12,6 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+from keras_cv.metrics.coco.coco_pymetric import COCOPyMetric
```
make a callback folder

tanzhenyu commented on keras_cv/metrics/coco/coco_pymetric.py:
```diff
@@ -0,0 +1,106 @@
+        logs.update(metrics)
+
+    def _eval_batch(self, batch, index):
+        images, gt_boxes = batch
```
This should be structured as 
images, gt_boxes, gt_classes = batch,
instead of gt_boxes [..., 5]
as the latter should be removed before 0.4

ianstenbit replied:
I agree that this should change, but I think it needs to change at the same time that we restructure this package-wide. Right now, for it to work as a callback it needs to accept data in the same format that the models do, which (at least for RetinaNet, which is what I'm testing with) is `images, boxes`, where boxes include the class.

I think this also may end up as something like
```
images, annotated_boxes = batch
boxes, classes = annotated_boxes
```
if we end up going with the nested concept of box classes (idk where we are on that conversation)

LukeWood replied:
+1 to waiting to restructure on a package wide basis.

tanzhenyu replied:
Hmmm, both the RetinaNet and FasterRCNN pipeline does not use any KPLs, so it's actually returning boxes and classes, and our model prediction returns boxes and classes separately, so I'm not sure what's blocking this?

tanzhenyu replied:
I believe we agreed that we should use : 
```
for examples in train_ds:
    image, gt_boxes, gt_classes = examples
```

This is the most native format for any detection workflow.

LukeWood replied:
Are we sure we want to diverge?  right now we have all bounding boxes in one format, and MaskRCNN in another format.

tanzhenyu commented on keras_cv/metrics/coco/coco_pymetric.py:
```diff
@@ -0,0 +1,106 @@
+        gt_classes = gt_boxes[:, :, 4]
+        gt_boxes = gt_boxes[:, :, :4]
+
+        source_ids = tf.strings.join(
```
I'd recommend taking the source id from the input dataset, this ideally should just be from "00032.jpg" -> 32

ianstenbit replied:
IMO this is a bad user experience, because it requires that the dataset has the source ID as a field. This isn't really exposed to the user in any meaningful way, it's just used internally by pycocoutils as a key for a row in the dataset.

If we require "source_id" as a field in the dataset, then users will have to manufacture a source ID themselves for datasets that don't already have one

tanzhenyu replied:
It is part of tfds, correct? https://www.tensorflow.org/datasets/catalog/coco
So in the future, we will probably provide some tools for users to find out bad predictions based on the metric output, it'd be nice to associate image id / filename with bad predictions

ianstenbit replied:
Yes, this is part of TFDS for COCO. However, using it here:
- requires users to parse that field and add it to the train/val datasets
- adds a lot of friction for using these metrics with any other dataset

Is there any value add to using it? From what Ic an see, we get the exact same behavior with a simpler API by leaving it out.

tanzhenyu replied:
Ok, that makes sense

tanzhenyu commented on keras_cv/metrics/coco/pycoco_utils.py:
```diff
@@ -0,0 +1,231 @@
+    metrics = coco_metrics
+
+    metrics_dict = {}
+    for i, name in enumerate(METRIC_NAMES):
```
do we need this given coco_eval.summarize() already gives the expected log?

ianstenbit replied:
Just verified this, and `coco_metrics` here is a list of numbers (without any associated metric names) -- so this is required to associated the metric names with the metrics for the output logs

tanzhenyu replied:
last time I tried it, it gives something like AP50, AP75, APSmall, APLarge, etc, is that not the case anymore?

ianstenbit replied:
coco_eval.eval gives a dictionary with e.g. `recall` and `precision`, but on a per-image basis. stats just gives an array. Implementation [here](https://github.com/cocodataset/cocoapi/blob/master/PythonAPI/pycocotools/cocoeval.py).

tanzhenyu commented on keras_cv/metrics/coco/pycoco_utils.py:
```diff
@@ -0,0 +1,231 @@
+
+import numpy as np
+import tensorflow as tf
+from pycocotools import coco
```
should do a try import and raise error and ask user to pip install

ianstenbit replied:
Good call -- done

tanzhenyu commented on keras_cv/metrics/coco/pycoco_utils.py:
```diff
@@ -0,0 +1,231 @@
+]
+
+
+class COCOWrapper(coco.COCO):
```
structure wise, let's make it:
1. keras_cv/metrics/pycoco_wrapper.py
2. keras_cv/callbacks/pycoco_callback.py
3. in the future if this becomes a PyMetrics, then keras_cv/metrics/pycoco_metrics.py

And we can make sure 1 can be re-used in 2 and 3 when implementing 3.

ianstenbit replied:
sgtm -- done

tanzhenyu reviewed (commented):
Thanks for the PR!

ianstenbit commented:
> * can we support evaluation for the train set from the callback too?

We can do this, yes. But I think the right approach for that is to support it in the Keras PyMetrics construct, which is forthcoming. For now, I think this should be eval-focused as this is a stopgap solution until we can build a proper PyMetric for this

ianstenbit commented:
> One high-level question is : we should probably first provide a VOCEvaluator given our existing training are all on Pascal VOC dataset. That would require some python level coding, but no need to rely on pycoco

I'm not sure I follow. This callback should work independently of which KerasCV model we're using for OD, as well as which dataset we're training it on.

LukeWood commented on keras_cv/metrics/coco/__init__.py:
```diff
@@ -13,5 +13,12 @@
-from keras_cv.metrics.coco.pycoco_wrapper import compute_pycoco_metrics
+
+try:
+    from keras_cv.metrics.coco.pycoco_wrapper import PyCOCOWrapper
```
This seems risky.  Can we instead do this on object instantiation?
```
def get_coco_eval():
    try:
        import cocoeval
        return cocoeval
     except e:
          raise ValueError(...)
```


then in the class:
```
cocoeval = get_coco_eval()
```

ianstenbit replied:
I don't think this will work, because we are defining a class which is a subclass of `COCO`.
Maybe there's a way to do this which I'm not aware of, but as far as I can tell, we can't define that class if the import fails.

LukeWood commented:
> One high-level question is : we should probably first provide a VOCEvaluator given our existing training are all on Pascal VOC dataset. That would require some python level coding, but no need to rely on pycoco

what would this look like?

tanzhenyu commented:
> > One high-level question is : we should probably first provide a VOCEvaluator given our existing training are all on Pascal VOC dataset. That would require some python level coding, but no need to rely on pycoco
> 
> I'm not sure I follow. This callback should work independently of which KerasCV model we're using for OD, as well as which dataset we're training it on.

COCO metrics is only for coco dataset. VOC metrics is defined slightly different and only need very few AP with fixed IOU threshold

tanzhenyu commented on keras_cv/callbacks/pycoco_callback.py:
```diff
@@ -0,0 +1,106 @@
+        logs.update(metrics)
+
+    def _eval_batch(self, batch, index):
+        images, gt_boxes = batch
```
I think we should only make sure this works with retinanet.fit and fasterrcnn.fit at this moment, so accepting boxes and classes separately is fine.

ianstenbit replied:
retinanet.fit assumes that boxes and classes are together in one tensor. This currently works with retinanet.fit but not fasterrcnn.fit. If I changed it, that would flip flop. More discussion in the thread above.

tanzhenyu replied:
Then it seems we should make sure this works with fasterrcnn.fit not retinanet.fit -- given we have discussed this several times and it seems that we have a consensus that we should separate them out, I'd prefer not introduce more tech debt.

ianstenbit replied:
Fine by me -- I will update this to work with fasterrcnn, not retinanet.

Once that's done, and I've added test, I'll re-request review.

LukeWood replied:
I believe we settled on a dictionary format, maybe it’s time migrate all components to the dictionary format for bounding boxes.

tanzhenyu reviewed (changes requested):
See comments inline

ianstenbit commented:
I've updated the callback to use the ```images, {"gt_boxes": boxes, "gt_classes": classes}``` approach for structuring input data, and I've added unit testing (using RetinaNet.fit for now, because FasterRCNN doesn't have a prediction decoder)

@tanzhenyu and @LukeWood PTAL -- and thank you both for your feedback so far!

ianstenbit commented:
BTW, test failures seem unrelated to this change and I am looking into it.

ianstenbit commented:
> BTW, test failures seem unrelated to this change and I am looking into it.

Turns out failures were due to this PR :) -- should be fixed now

LukeWood commented on keras_cv/callbacks/pycoco_callback.py:
```diff
@@ -0,0 +1,102 @@
+
+
+class PyCOCOCallback(Callback):
+    def __init__(self, validation_data, bounding_box_format):
```
add **kwargs and propogate to __init__

ianstenbit replied:
Done

tanzhenyu reviewed (approved):
LGTM, only need to make sure:
1) it's using the input format as (images, {"gt_boxes", "gt_classes"})
2) having a description on the PR on the eval results (for example, 3 epochs)

tanzhenyu commented:
/gcbrun

ianstenbit commented:
> LGTM, only need to make sure:
> 
> 1. it's using the input format as (images, {"gt_boxes", "gt_classes"})
> 2. having a description on the PR on the eval results (for example, 3 epochs)

SG -- I am working on getting the eval results. I figured I'd get it working on TPU at the same time, so it's taking a bit. I'll wait to merge until that is sorted out.

ianstenbit commented:
I've updated the description with a TensorBoard link to an example run training a RetinaNet with this callback.

I think there are some improvements to be made on the performance which are likely going to be tied to more efficiently getting predictions from the model during evaluation time (calling the model for each batch instead of for the full dataset is super inefficient). I will work on those improvements as a follow-up PR.

ianstenbit commented:
/gcbrun

tanzhenyu commented:
> https://tensorboard.dev/experiment/Dmvt2HYDTxamopnh5QgJkQ/#scalars&_smoothingWeight=0&runSelectionState=eyJ0cmFpbiI6dHJ1ZX0%3D

Sounds good to me. Feel free to go ahead and merge, with follow-up PRs
